In [1]:
import sqlite3
import psycopg2
from sqlalchemy import create_engine

In [2]:
import pandas as pd

In [3]:
conn = sqlite3.connect('ladder.db')

In [4]:
conn

### Simple SELECTs

Query all the data in the pets table.

In [5]:
pd.read_sql("SELECT * FROM pets;", con = conn)

name sex  species  age
0        Chloe   F      dog    9
1   Paddington   M      dog   10
2        Petey   M      dog    7
3        Bella   F      dog    1
4   Glenn Coco   M      cat    6
5       Alanna   F      cat    3
6         Mimi   F      dog    4
7        Midge   F      cat    7
8          Eli   M      dog    8
9        Shuri   F      cat    2
10      Oliver   M      cat    5
11      Dottie   F      cat    3
12      Pinchy   M  lobster    3

Query only the first 5 rows of the pets table.

In [6]:
pd.read_sql("""SELECT * FROM pets
                LIMIT 5;""", con = conn)

name sex species  age
0       Chloe   F     dog    9
1  Paddington   M     dog   10
2       Petey   M     dog    7
3       Bella   F     dog    1
4  Glenn Coco   M     cat    6

Query only the names and ages of the pets in the pets table.

In [7]:
pd.read_sql("""SELECT name,
            age
            FROM pets;""", con=conn)

name  age
0        Chloe    9
1   Paddington   10
2        Petey    7
3        Bella    1
4   Glenn Coco    6
5       Alanna    3
6         Mimi    4
7        Midge    7
8          Eli    8
9        Shuri    2
10      Oliver    5
11      Dottie    3
12      Pinchy    3

Query the pets in the pets table, sorted youngest to oldest

In [8]:
pd.read_sql("""SELECT * FROM pets
                ORDER BY age ASC;""", con = conn)

name sex  species  age
0        Bella   F      dog    1
1        Shuri   F      cat    2
2       Alanna   F      cat    3
3       Dottie   F      cat    3
4       Pinchy   M  lobster    3
5         Mimi   F      dog    4
6       Oliver   M      cat    5
7   Glenn Coco   M      cat    6
8        Petey   M      dog    7
9        Midge   F      cat    7
10         Eli   M      dog    8
11       Chloe   F      dog    9
12  Paddington   M      dog   10

Query the pets in the pets table alphabetically.

In [9]:
pd.read_sql("""SELECT * FROM pets
                ORDER BY name ASC;""", con = conn)

name sex  species  age
0       Alanna   F      cat    3
1        Bella   F      dog    1
2        Chloe   F      dog    9
3       Dottie   F      cat    3
4          Eli   M      dog    8
5   Glenn Coco   M      cat    6
6        Midge   F      cat    7
7         Mimi   F      dog    4
8       Oliver   M      cat    5
9   Paddington   M      dog   10
10       Petey   M      dog    7
11      Pinchy   M  lobster    3
12       Shuri   F      cat    2

Query all the male pets in the pets table

In [10]:
pd.read_sql("""SELECT * FROM pets
                WHERE sex = "M";""", con = conn)

name sex  species  age
0  Paddington   M      dog   10
1       Petey   M      dog    7
2  Glenn Coco   M      cat    6
3         Eli   M      dog    8
4      Oliver   M      cat    5
5      Pinchy   M  lobster    3

Query all the cats in the pets table.

In [11]:
pd.read_sql("""SELECT * FROM pets
                WHERE species = "cat";""", con = conn)

name sex species  age
0  Glenn Coco   M     cat    6
1      Alanna   F     cat    3
2       Midge   F     cat    7
3       Shuri   F     cat    2
4      Oliver   M     cat    5
5      Dottie   F     cat    3

Query all the pets in the pets table that are at least 5 years old

In [12]:
pd.read_sql("""SELECT * FROM pets
                WHERE age >= 5
                ORDER BY age ASC;""", con = conn)

name sex species  age
0      Oliver   M     cat    5
1  Glenn Coco   M     cat    6
2       Petey   M     dog    7
3       Midge   F     cat    7
4         Eli   M     dog    8
5       Chloe   F     dog    9
6  Paddington   M     dog   10

Query all the male dogs in the pets table. Do not include the sex or species column, since you already know them.

In [13]:
pd.read_sql("""SELECT name,
                age,
                species,
                sex
                FROM pets
                WHERE sex = "M" 
                AND species = "dog";""", con = conn)


name  age species sex
0  Paddington   10     dog   M
1       Petey    7     dog   M
2         Eli    8     dog   M

Get all the names of the dogs in the pets table that are younger than 5 years old

In [14]:
pd.read_sql("""SELECT name,
                age
                FROM pets
                WHERE age < 5
                ORDER BY age ASC;""", con = conn)


name  age
0   Bella    1
1   Shuri    2
2  Alanna    3
3  Dottie    3
4  Pinchy    3
5    Mimi    4

Query all the pets in the pets table that are either male dogs or female cats.

In [15]:
pd.read_sql("""SELECT *
                FROM pets
                WHERE (species = "dog" AND sex = "M")
                OR (species = "cat" AND sex = "F")
                ORDER BY species, age ASC;""", con = conn)

name sex species  age
0       Shuri   F     cat    2
1      Alanna   F     cat    3
2      Dottie   F     cat    3
3       Midge   F     cat    7
4       Petey   M     dog    7
5         Eli   M     dog    8
6  Paddington   M     dog   10

Query the five oldest pets in the pets table.

In [16]:
pd.read_sql("""SELECT *
                FROM pets
                ORDER by age DESC
                LIMIT 5;""", con = conn)

name sex species  age
0  Paddington   M     dog   10
1       Chloe   F     dog    9
2         Eli   M     dog    8
3       Petey   M     dog    7
4       Midge   F     cat    7

Get the names and ages of all the female cats in the pets table sorted by age, descending.


In [17]:
pd.read_sql("""SELECT name,
                      age
                      FROM pets
                      WHERE species = "cat" AND sex = "F"
                      ORDER BY age DESC;""", con = conn)

name  age
0   Midge    7
1  Alanna    3
2  Dottie    3
3   Shuri    2

Get all pets from pets whose names start with P.


In [18]:
pd.read_sql("""SELECT *
                FROM pets
                where name like "P%";""", con = conn)

name sex  species  age
0  Paddington   M      dog   10
1       Petey   M      dog    7
2      Pinchy   M  lobster    3

Select all employees from employees_null where the salary is missing.

In [19]:
pd.read_sql("""SELECT *
                FROM employees_null
                WHERE salary IS NULL""", con = conn)

ID firstname lastname            job salary   startdate
0  22     Maria     Owen          Sales   None  2010-03-22
1  47   Shannon   Rivera  Administrator   None  2007-09-12
2  53  Jennifer     Cruz          Sales   None  2015-11-09
3  56   Russell     Rice          Sales   None  2008-09-02
4  59     Andre   Levine          Sales   None  2016-03-27
5  60      Anna  Fischer          Sales   None  1994-12-25
6  69  Danielle      Orr          Sales   None  2004-06-24
7  79    Tamara  Douglas          Sales   None  2017-10-03
8  85   Candice   Wright          Sales   None  1998-03-04
9  98   Bradley   Romero  Administrator   None  2009-05-10

Select all employees from employees_null where the salary is below $35,000 or missing.

In [20]:
pd.read_sql("""SELECT *
                FROM employees_null
                WHERE salary < 35000 OR salary IS NULL;""", con = conn)

ID firstname  lastname            job   salary   startdate
0    22     Maria      Owen          Sales      NaN  2010-03-22
1    26  Samantha   Nichols          Sales  33248.0  2015-04-20
2    47   Shannon    Rivera  Administrator      NaN  2007-09-12
3    53  Jennifer      Cruz          Sales      NaN  2015-11-09
4    55    Julian  Martinez          Sales  34597.0  2009-06-19
5    56   Russell      Rice          Sales      NaN  2008-09-02
6    59     Andre    Levine          Sales      NaN  2016-03-27
7    60      Anna   Fischer          Sales      NaN  1994-12-25
8    65   Michael      West          Sales  32781.0  1991-08-08
9    69  Danielle       Orr          Sales      NaN  2004-06-24
10   73    Miguel  Williams          Sales  31584.0  2004-04-02
11   79    Tamara   Douglas          Sales      NaN  2017-10-03
12   85   Candice    Wright          Sales      NaN  1998-03-04
13   98   Bradley    Romero  Administrator      NaN  2009-05-10
14  101   UNKNOWN   UNKNOWN           None     -1.0        None

Select all employees from employees_null where the job title is missing. What do you see?

This is likely an incorrect record because the basic identifying information is not available.

In [21]:
pd.read_sql("""SELECT *
                FROM employees_null
                WHERE JOB IS NULL;""", con = conn)

ID firstname lastname   job  salary startdate
0  101   UNKNOWN  UNKNOWN  None      -1      None

Who is the newest employee in employees? The most senior?

The newest employee is Roger Conner and the most senior employee is Mary Nash.

In [22]:
pd.read_sql("""SELECT *
                FROM employees_null
                ORDER BY startdate DESC
                LIMIT 1;""", con = conn)

ID firstname lastname    job  salary   startdate
0  19     Roger   Conner  Sales   60375  2019-05-24

In [23]:
pd.read_sql("""SELECT *
                FROM employees_null
                WHERE startdate IS NOT NULL
                ORDER BY startdate ASC
                LIMIT 1;""", con = conn)

ID firstname lastname    job  salary   startdate
0  32      Mary     Nash  Sales   65472  1990-02-16

Select all employees from employees named Thomas.

In [24]:
pd.read_sql("""SELECT *
                FROM employees_null
                WHERE firstname = "Thomas" """, con = conn)

ID firstname  lastname    job  salary   startdate
0   1    Thomas      Peck  Sales  112972  2011-01-04
1  45    Thomas    Miller  Sales   44836  2002-11-17
2  80    Thomas  Gonzalez  Sales   84772  2019-01-05

Select all employees from employees named Thomas or Shannon.

In [25]:
pd.read_sql("""SELECT *
                FROM employees_null
                WHERE firstname = "Thomas" OR firstname = "Shannon" """, con = conn)

ID firstname  lastname            job    salary   startdate
0   1    Thomas      Peck          Sales  112972.0  2011-01-04
1  33   Shannon    Bailey          Sales   66071.0  2014-12-30
2  45    Thomas    Miller          Sales   44836.0  2002-11-17
3  47   Shannon    Rivera  Administrator       NaN  2007-09-12
4  80    Thomas  Gonzalez          Sales   84772.0  2019-01-05

Select all employees from employees named Robert, Lisa, or any name that begins with a J. In addition, only show employees who are not in sales. This will be a little bit of a longer query.

In [26]:
pd.read_sql("""SELECT *
                FROM employees_null
                WHERE (firstname = "Robert" 
                OR firstname = "Lisa"
                OR firstname like "J%")
                AND job NOT IN('Sales');""", con = conn)

ID firstname lastname            job  salary   startdate
0   4    Janice   Martin             IT   62007  2011-02-25
1   6      Lisa   Cooper     Operations   56265  1994-02-20
2   8      Jose  Rosario  Administrator  111905  2008-05-05
3  39     Jenny   Barber             IT  110679  2019-01-10
4  72    Joseph   Campos     Operations   88345  1992-06-10
5  78    Robert     Diaz             IT   37397  2016-12-29

### Column Operations

Query the top 5 rows of the employees table to get a glimpse of these new data.

In [27]:
sql = """
    SELECT *
    FROM employees
    LIMIT 5;"""

pd.read_sql(sql, con=conn)

ID    firstname  lastname    job  salary   startdate
0   0    Christine  Thompson  Sales  123696  2005-01-20
1   1       Thomas      Peck  Sales  112972  2011-01-04
2   2  Christopher    Robles     IT   78426  2003-12-10
3   3    Elizabeth     Munoz  Sales   55824  1993-07-28
4   4       Janice    Martin     IT   62007  2011-02-25

Query the employees table, but convert their salaries to Euros.

Repeat the previous problem, but rename the column salary_eu.

In [28]:
sql = """
    SELECT ID,
    firstname,
    lastname,
    job,
    salary,
    ROUND(salary * 1.1,1) AS salary_eu,
    startdate
    FROM employees
    LIMIT 5;"""

pd.read_sql(sql, con=conn)

ID    firstname  lastname    job  salary  salary_eu   startdate
0   0    Christine  Thompson  Sales  123696   136065.6  2005-01-20
1   1       Thomas      Peck  Sales  112972   124269.2  2011-01-04
2   2  Christopher    Robles     IT   78426    86268.6  2003-12-10
3   3    Elizabeth     Munoz  Sales   55824    61406.4  1993-07-28
4   4       Janice    Martin     IT   62007    68207.7  2011-02-25

Query the employees table, but combine the firstname and lastname columns to be "Firstname, Lastname" format. Call this column fullname. For example, the first row should contain Thompson, Christine as fullname. Also, display the rounded salary_eu instead of salary.

In [29]:
sql = """
    SELECT ID,
    firstname,
    lastname,
    lastname ||", "|| firstname AS fullname,
    job,
    ROUND(salary * 1.1, 1) AS salary_eu,
    startdate
    FROM employees
    LIMIT 5;"""

pd.read_sql(sql, con=conn)

ID    firstname  lastname             fullname    job  salary_eu  \
0   0    Christine  Thompson  Thompson, Christine  Sales   136065.6   
1   1       Thomas      Peck         Peck, Thomas  Sales   124269.2   
2   2  Christopher    Robles  Robles, Christopher     IT    86268.6   
3   3    Elizabeth     Munoz     Munoz, Elizabeth  Sales    61406.4   
4   4       Janice    Martin       Martin, Janice     IT    68207.7   

    startdate  
0  2005-01-20  
1  2011-01-04  
2  2003-12-10  
3  1993-07-28  
4  2011-02-25

Query the employees table, but replace startdate with startyear using the SUBSTR() function. Also include fullname and salary_eu

In [30]:
sql = """
    SELECT ID,
    firstname,
    lastname,
    lastname ||", "|| firstname AS fullname,
    job,
    ROUND(salary * 1.1, 1) AS salary_eu,
    startdate,
    SUBSTR(startdate, 1, 4) AS start_year
    FROM employees
    LIMIT 5;"""

pd.read_sql(sql, con=conn)

ID    firstname  lastname             fullname    job  salary_eu  \
0   0    Christine  Thompson  Thompson, Christine  Sales   136065.6   
1   1       Thomas      Peck         Peck, Thomas  Sales   124269.2   
2   2  Christopher    Robles  Robles, Christopher     IT    86268.6   
3   3    Elizabeth     Munoz     Munoz, Elizabeth  Sales    61406.4   
4   4       Janice    Martin       Martin, Janice     IT    68207.7   

    startdate start_year  
0  2005-01-20       2005  
1  2011-01-04       2011  
2  2003-12-10       2003  
3  1993-07-28       1993  
4  2011-02-25       2011

Repeat the above problem, but instead of using SUBSTR(), use STRFTIME()

In [31]:
sql = """
    SELECT ID,
    firstname,
    lastname,
    lastname ||", "|| firstname AS fullname,
    job,
    ROUND(salary * 1.1, 1) AS salary_eu,
    STRFTIME('%Y',startdate) AS start_year
    FROM employees
    LIMIT 5;"""

pd.read_sql(sql, con=conn)

ID    firstname  lastname             fullname    job  salary_eu start_year
0   0    Christine  Thompson  Thompson, Christine  Sales   136065.6       2005
1   1       Thomas      Peck         Peck, Thomas  Sales   124269.2       2011
2   2  Christopher    Robles  Robles, Christopher     IT    86268.6       2003
3   3    Elizabeth     Munoz     Munoz, Elizabeth  Sales    61406.4       1993
4   4       Janice    Martin       Martin, Janice     IT    68207.7       2011

Query the employees table, replacing firstname/lastname with fullname and startdate with startyear. Print out the salary in USD again, except format it with a dollar sign, comma separators, and no decimal. For example, the first row should read $123,696. This column should still be named salary.

In [32]:
sql = """
    SELECT ID,
    firstname,
    lastname,
    lastname ||", "|| firstname AS fullname,
    job,
    STRFTIME('%Y',startdate) AS start_year,
    PRINTF("$%,.2d",salary) AS salary
    FROM employees
    LIMIT 5;"""

pd.read_sql(sql, con=conn)

ID    firstname  lastname             fullname    job start_year    salary
0   0    Christine  Thompson  Thompson, Christine  Sales       2005  $123,696
1   1       Thomas      Peck         Peck, Thomas  Sales       2011  $112,972
2   2  Christopher    Robles  Robles, Christopher     IT       2003   $78,426
3   3    Elizabeth     Munoz     Munoz, Elizabeth  Sales       1993   $55,824
4   4       Janice    Martin       Martin, Janice     IT       2011   $62,007

Last year, only salespeople were eligible for bonuses. Create a column bonus that is "Yes" if you're eligible for a bonus, otherwise "No"

In [33]:
sql = """
    SELECT *,
        (CASE WHEN job = "Sales" THEN "Yes" ELSE "No" END) AS bonuses
    FROM employees
    LIMIT 5;"""

pd.read_sql(sql, con=conn)

ID    firstname  lastname    job  salary   startdate bonuses
0   0    Christine  Thompson  Sales  123696  2005-01-20     Yes
1   1       Thomas      Peck  Sales  112972  2011-01-04     Yes
2   2  Christopher    Robles     IT   78426  2003-12-10      No
3   3    Elizabeth     Munoz  Sales   55824  1993-07-28     Yes
4   4       Janice    Martin     IT   62007  2011-02-25      No

This year, only sales people with a salary of 100000 or higher are eligible for bonuses. Create a bonus column like in the last problem for salespeople with salaries at least $100,000.

In [34]:
sql = """
    SELECT *,
        (CASE WHEN job = "Sales" AND salary >= 100000 THEN "Yes" ELSE "No" END) AS bonus
        FROM employees
        LIMIT 5;"""

pd.read_sql(sql, con=conn)

ID    firstname  lastname    job  salary   startdate bonus
0   0    Christine  Thompson  Sales  123696  2005-01-20   Yes
1   1       Thomas      Peck  Sales  112972  2011-01-04   Yes
2   2  Christopher    Robles     IT   78426  2003-12-10    No
3   3    Elizabeth     Munoz  Sales   55824  1993-07-28    No
4   4       Janice    Martin     IT   62007  2011-02-25    No

Next year, the bonus structure will be a little more complicated. You'll create a target_comp column which represents an employee's target total compensation after their bonus. Here is the company's bonus structure:

* Salespeople who make more than 100,000 will be eligible for a 10% bonus.
* Salespeople who make less than 100,000 will be eligible for a 5% bonus.
* Administrators will also be eligible for a 5% bonus.

Anyone who does not meet any of the above descriptions is not eligible for a bonus.

Create this target_comp column, making sure to format both the salary and target_comp columns nicely (ie, with dollar 
signs and comma separators).

In [35]:
sql = """
    SELECT ID,
        firstname,
        lastname,
        job,
        PRINTF("$%,.2d",salary) AS salary,
        PRINTF("$%,.2d",(CASE WHEN job = "Sales" AND salary >= 100000 THEN 0.10
            WHEN job = "Sales" AND salary < 100000 THEN 0.05
            WHEN job = "Administrator" THEN 0.05
            WHEN job NOT IN('Sales','Administrator') THEN 0
        END) * salary + salary) AS target_comp,
        startdate
        FROM employees
        LIMIT 5;"""

pd.read_sql(sql, con=conn)

ID    firstname  lastname    job    salary target_comp   startdate
0   0    Christine  Thompson  Sales  $123,696    $136,065  2005-01-20
1   1       Thomas      Peck  Sales  $112,972    $124,269  2011-01-04
2   2  Christopher    Robles     IT   $78,426     $78,426  2003-12-10
3   3    Elizabeth     Munoz  Sales   $55,824     $58,615  1993-07-28
4   4       Janice    Martin     IT   $62,007     $62,007  2011-02-25

In [36]:
sql = """SELECT job,COUNT(*)
            FROM employees
            GROUP BY job"""

pd.read_sql(sql, con=conn)

job  COUNT(*)
0  Administrator        14
1             IT        10
2     Operations         8
3          Sales        68

## Summarizing Data with SQL

### Summary Statistics

How many rows are in the pets table?

In [37]:
sql = """SELECT COUNT(*) AS number_of_rows
        FROM pets;"""

pd.read_sql(sql, con=conn)

number_of_rows
0              13

How many female pets are in the pets table?

In [38]:
sql = """SELECT COUNT(*) AS nbr_female
         FROM pets
         WHERE sex = "F";"""

pd.read_sql(sql, con=conn)

nbr_female
0           7

How many female cats are in the pets table?

In [39]:
sql = """SELECT COUNT(*) AS nbr_female_cat
         FROM pets
         WHERE sex = "F"
         AND species = "cat";"""

pd.read_sql(sql, con=conn)

nbr_female_cat
0               4

What's the mean age of pets in the pets table?

In [40]:
sql = """SELECT AVG(age) AS avg_pet_age
         FROM pets;"""

pd.read_sql(sql, con=conn)

avg_pet_age
0     5.230769

What's the mean age of dogs in the pets table?

In [41]:
sql = """SELECT AVG(age) AS avg_dog_age
         FROM pets
         WHERE species = "dog";"""

pd.read_sql(sql, con=conn)

avg_dog_age
0          6.5

What's the mean age of male dogs in the pets table?

In [42]:
sql = """SELECT AVG(age) AS avg_male_dog_age
         FROM pets
         WHERE species = "dog"
         AND sex = "M";"""

pd.read_sql(sql, con=conn)

avg_male_dog_age
0          8.333333

What's the count, mean, minimum, and maximum of pet ages in the pets table?

In [43]:
sql = """SELECT COUNT(age) AS "Count of Age",
            ROUND(AVG(age),1) AS "Average Age",
            MIN(age) AS "Minimum Age",
            MAX(age) AS "Maximum Age"
         FROM pets"""

pd.read_sql(sql, con=conn)

Count of Age  Average Age  Minimum Age  Maximum Age
0            13          5.2            1           10

How many rows in employees_null have missing salaries

In [44]:
sql = """SELECT COUNT(*)
         FROM employees_null
         WHERE salary IS NULL"""

pd.read_sql(sql, con=conn)

COUNT(*)
0        10

How many salespeople in employees_null having nonmissing salaries?

In [45]:
sql = """SELECT COUNT(*)
         FROM employees_null
         WHERE salary IS NOT NULL"""

pd.read_sql(sql, con=conn)

COUNT(*)
0        91

What's the mean salary of employees who joined the company after 2010? Go back to the usual employees table for this one

In [46]:
sql = """SELECT AVG(salary) AS avg_salary
         FROM employees
         WHERE CAST(SUBSTR(startdate, 1, 4) AS REAL) > 2010;"""

pd.read_sql(sql, con=conn)


avg_salary
0  78554.088235

What's the mean salary of employees in Swiss Francs?
* Hint: Swiss Francs are abbreviated "CHF" and 1 USD = 0.97 CHF.

In [47]:
sql = """SELECT AVG(salary) * 0.97 AS avg_salary_chf
         FROM employees;"""

pd.read_sql(sql, con=conn)

avg_salary_chf
0      75727.5605

Create a query that computes the mean salary in USD as well as CHF. Give the columns human-readable names (for example "Mean Salary in USD"). Also, format them with comma delimiters and currency symbols.

* NOTE: Comma-delimiting numbers is only available for integers in SQLite, so rounding (down) to the nearest dollar or franc will be done for us.
* NOTE2: The symbols for francs is simply Fr. or fr.. So an example output will look like 100,000 Fr.

In [48]:
sql = """SELECT PRINTF("$%,.2d", AVG(salary) * 0.97) || " Fr." AS "Mean Salary in CHF",
            PRINTF("$%,.2d",AVG(salary)) || " USD" AS "Mean Salary in USD"
         FROM employees;"""

pd.read_sql(sql, con=conn)


Mean Salary in CHF Mean Salary in USD
0        $75,727 Fr.        $78,069 USD

### Aggregating Statistics with GROUP BY

What is the average age of pets by species?

Repeat the previous problem but make sure the species label is also displayed! Assume this behavior is always being asked of you any time you use GROUP BY

In [49]:
sql = """SELECT species,ROUND(AVG(age),0) AS "Average Age"
        FROM pets
        GROUP BY species;"""

pd.read_sql(sql, con=conn)

species  Average Age
0      cat          4.0
1      dog          7.0
2  lobster          3.0

What is the count, mean, minimum, and maximum age by species in pets?

In [50]:
sql = """SELECT species,
            COUNT(age),
            AVG(age),
            MIN(age),
            MAX(age)
         FROM pets
         GROUP BY species;"""

pd.read_sql(sql, con=conn)

species  COUNT(age)  AVG(age)  MIN(age)  MAX(age)
0      cat           6  4.333333         2         7
1      dog           6  6.500000         1        10
2  lobster           1  3.000000         3         3

Show the mean salaries of each job title in employees

In [51]:
sql = """SELECT job,AVG(salary)
         FROM employees
         GROUP BY job;"""

pd.read_sql(sql, con=conn)

job   AVG(salary)
0  Administrator  71986.142857
1             IT  71381.000000
2     Operations  74055.250000
3          Sales  80778.044118

Show the mean salaries in New Zealand dollars of each job title in employees
* NOTE: 1 USD = 1.65 NZD

In [52]:
sql = """SELECT job,PRINTF("$%,.2d",AVG(salary) * 1.65) || " NSD" AS "Mean Salaries in NZD"
         FROM employees
         GROUP BY job;"""

pd.read_sql(sql, con=conn)

job Mean Salaries in NZD
0  Administrator         $118,777 NSD
1             IT         $117,778 NSD
2     Operations         $122,191 NSD
3          Sales         $133,283 NSD

Show the mean, min, and max salaries of each job title in employees, as well as the numbers of employees in each category.

In [53]:
sql = """SELECT job,
            COUNT(*),
            AVG(salary),
            MIN(salary),
            MAX(salary)
         FROM employees
         GROUP BY job;"""

pd.read_sql(sql, con=conn)

job  COUNT(*)   AVG(salary)  MIN(salary)  MAX(salary)
0  Administrator        14  71986.142857        41151       120492
1             IT        10  71381.000000        37397       115729
2     Operations         8  74055.250000        41797       108989
3          Sales        68  80778.044118        31333       124474

Show the mean salaries of each job title in employees sorted descending by salary

In [54]:
sql = """SELECT job, AVG(salary)
             FROM employees
             GROUP BY job
             ORDER BY AVG(salary) DESC;"""

pd.read_sql(sql, con=conn)

job   AVG(salary)
0          Sales  80778.044118
1     Operations  74055.250000
2  Administrator  71986.142857
3             IT  71381.000000

What are the top 5 most common first names among employees?

In [55]:
sql = """SELECT firstname,COUNT(*)
            FROM employees
            GROUP BY firstname
            ORDER BY COUNT(*) DESC;"""

pd.read_sql(sql, con=conn)

firstname  COUNT(*)
0      Thomas         3
1      Robert         3
2     Michael         3
3        Lisa         3
4     William         2
..        ...       ...
78     Amanda         1
79      Allen         1
80  Alexander         1
81  Alejandro         1
82      Aaron         1

[83 rows x 2 columns]

Show all first names which have exactly 2 occurrences in employees
* Use HAVING as a way to add additional filter when using aggregate functions

In [56]:
sql = """SELECT firstname,COUNT(*) AS cnt_firstname
            FROM employees
            GROUP BY firstname
            HAVING cnt_firstname = 2
            ORDER BY COUNT(*) DESC;"""

pd.read_sql(sql, con=conn) 

firstname  cnt_firstname
0      William              2
1      Shannon              2
2         Mark              2
3       Leslie              2
4       Joseph              2
5        Jacob              2
6    Elizabeth              2
7       Donald              2
8  Christopher              2

Take a look at the transactions table to get a idea of what it contains. Note that a transaction may span multiple rows if different items are purchased as part of the same order. The employee who made the order is also given by their ID.

In [57]:
sql = """SELECT *
         FROM transactions
         LIMIT 10;"""

pd.read_sql(sql, con=conn)

order_id                   customer  unit_price  quantity   orderdate  \
0         0  Kent, Curtis and Williams       20.50        12  2010-05-30   
1         0  Kent, Curtis and Williams       24.00        11  2010-05-30   
2         0  Kent, Curtis and Williams       22.25        14  2010-05-30   
3         0  Kent, Curtis and Williams       10.50        11  2010-05-30   
4         0  Kent, Curtis and Williams       13.75         9  2010-05-30   
5         0  Kent, Curtis and Williams       18.00        11  2010-05-30   
6         1          Matthews and Sons       18.25         4  2005-12-04   
7         1          Matthews and Sons       18.25        11  2005-12-04   
8         1          Matthews and Sons        7.50         8  2005-12-04   
9         1          Matthews and Sons       23.25         5  2005-12-04   

   employee_id  
0           81  
1           81  
2           81  
3           81  
4           81  
5           81  
6            0  
7            0  
8            0  
9            0

Show the top 5 largest orders (and their respective customer) in terms of the numbers of items purchased in that order.

In [58]:
sql = """SELECT customer,
            quantity
        FROM transactions
        ORDER BY quantity DESC
        LIMIT 5;"""

pd.read_sql(sql, con=conn)

customer  quantity
0  Turner, Horn and Adams        27
1                Hart LLC        25
2           Torres-Murray        24
3               Hart-Lynn        24
4        Sandoval-Jackson        24

Show the total cost of each transaction.
* Hint: The unit_price column is the price of one item. The customer may have purchased multiple.
* Hint2: Note that transactions here span multiple rows if different items are purchased.

Show the top 5 transactions in terms of total cost.

In [59]:
sql = """SELECT order_id, 
            SUM(quantity * unit_price) AS "total_price"
         FROM transactions
         GROUP BY order_id
         ORDER BY "total_price" DESC
         LIMIT 5;"""

pd.read_sql(sql, con=conn)

order_id  total_price
0      2190      2369.75
1      6324      2333.50
2      3009      2317.25
3      4726      2300.25
4      2415      2297.75

Show the top 5 customers in terms of total revenue (ie, which customers have we done the most business with in terms of money?)

In [60]:
sql = """SELECT customer,
            PRINTF("$%,.2d",SUM(quantity * unit_price)) AS "total_revenue"
         FROM transactions
         GROUP BY customer
         ORDER by total_revenue DESC
         LIMIT 5;"""

pd.read_sql(sql, con=conn)

customer total_revenue
0  Turner, Horn and Adams       $89,645
1              Reeves Inc       $85,485
2           Lara-Robinson       $84,383
3             Evans-Parks       $83,294
4          Watson-Harrell       $81,818

Show the top 5 employees in terms of revenue generated (ie, which employees made the most in sales?)

In [61]:
sql = """SELECT employee_id,
            PRINTF("$%,.2d",SUM(unit_price * quantity)) AS "total_revenue_by_employee"
         FROM transactions
         GROUP BY employee_id
         ORDER BY "total_revenue_by_employee" DESC
         LIMIT 5;"""

pd.read_sql(sql, con=conn)

employee_id total_revenue_by_employee
0           25                   $99,349
1           96                   $98,962
2           59                   $98,131
3           34                   $98,102
4           79                   $97,537

Show all customers who've done more than $80,000 worth of business with us

In [62]:
sql = """SELECT customer,
            PRINTF("$%,.2d",SUM(quantity * unit_price)) AS "total_revenue"
         FROM transactions
         GROUP BY customer
         HAVING SUM(quantity * unit_price) > 80000
         ORDER by total_revenue DESC;"""

pd.read_sql(sql, con=conn)

customer total_revenue
0  Turner, Horn and Adams       $89,645
1              Reeves Inc       $85,485
2           Lara-Robinson       $84,383
3             Evans-Parks       $83,294
4          Watson-Harrell       $81,818
5            Ellis-Miller       $81,595
6                Hart LLC       $80,331
7          Williams-Jones       $80,152

Which customer worked with the largest number of employees?

In [63]:
sql = """SELECT customer, 
            COUNT(DISTINCT employee_id) AS "nbr_employees"
         FROM transactions
         GROUP BY customer
         ORDER BY "nbr_employees" DESC
         LIMIT 1;"""
         
pd.read_sql(sql, con=conn)

customer  nbr_employees
0  Turner, Horn and Adams             61

### Views, CTEs, and Joins

62) Look at the `yum` table. It is the stock data for Yum! Brands, Inc. from 2015 through 2019. Yum! is the company that owns Taco Bell, the best restaurant.


In [64]:
sql = """SELECT *
         FROM yum
         LIMIT 5;"""

pd.read_sql(sql, con = conn)

date       open       high        low      close   volume
0  2015-01-02  52.796551  52.846874  51.761322  52.012939  2283400
1  2015-01-05  52.056076  52.278934  50.948959  50.956146  4418600
2  2015-01-06  51.301224  51.329979  49.877785  50.330696  5004400
3  2015-01-07  51.020847  52.056076  50.884254  51.998562  4554100
4  2015-01-08  52.408340  53.040977  52.365204  52.904385  4258200

63) Query the `yum` table, aggregating by **both** month and year, with the following resulting columns:
* Year (4 digits)
* Month
* Average open, high, low, and close
* Total volume

Finally, sort this data so it's in proper chronological order.

In [65]:
sql = """SELECT SUBSTR(date,1,4) AS year,
            SUBSTR(date,6,2) AS month,
            AVG(open) AS avg_open,
            AVG(high) AS avg_high,
            AVG(low) AS avg_low,
            AVG(close) AS avg_close,
            SUM(volume) AS total_volume
         FROM yum
         GROUP BY year, month
         LIMIT 12;"""

pd.read_sql(sql, con = conn)

year month   avg_open   avg_high    avg_low  avg_close  total_volume
0   2015    01  52.221423  52.736520  51.694464  52.192667      89074400
1   2015    02  54.183662  54.823868  53.723561  54.534791      98621800
2   2015    03  56.845958  57.324685  56.445657  56.818509     108827600
3   2015    04  58.533429  59.140050  58.168498  58.740885     117743300
4   2015    05  65.739757  66.475916  65.144141  65.755572     131485600
5   2015    06  65.550291  66.017907  65.188876  65.579373      69412500
6   2015    07  64.009869  64.498399  63.262531  63.883079     116786700
7   2015    08  60.153709  60.950327  59.170176  60.029784     155692800
8   2015    09  57.495806  58.036356  56.810106  57.506419      97615900
9   2015    10  52.663224  53.319391  51.934187  52.669760     232475400
10  2015    11  50.964415  51.363767  50.513300  50.978792      98766200
11  2015    12  52.933142  53.467747  52.325012  52.947193     116614800

64) Save the results of the previous query as a view named `yum_by_month`.

In [66]:
sql = """CREATE VIEW yum_by_month_v
         AS
         SELECT SUBSTR(date,1,4) AS year,
            SUBSTR(date,6,2) AS month,
            AVG(open) AS avg_open,
            AVG(high) AS avg_high,
            AVG(low) AS avg_low,
            AVG(close) AS avg_close,
            SUM(volume) AS total_volume
         FROM yum
         GROUP BY year, month;"""
### Created in sqlite3

In [67]:
sql = """SELECT *
         FROM yum_by_month_v
         LIMIT 5;"""

pd.read_sql(sql, con=conn)

year month   avg_open   avg_high    avg_low  avg_close  total_volume
0  2015    01  52.221423  52.736520  51.694464  52.192667      89074400
1  2015    02  54.183662  54.823868  53.723561  54.534791      98621800
2  2015    03  56.845958  57.324685  56.445657  56.818509     108827600
3  2015    04  58.533429  59.140050  58.168498  58.740885     117743300
4  2015    05  65.739757  66.475916  65.144141  65.755572     131485600

65) Create a view of `transactions` consisting of only three columns: year, month, and total sales in that month. Call this view `trans_by_month`.

In [68]:
sql = """SELECT SUBSTR(date,1,4) AS year,
            SUBSTR(date,6,2) AS month,
            SUM(volume) AS total_sales
         FROM yum
         GROUP BY year,month
         LIMIT 5;"""

66) Create a view of `transactions` consisting of only two columns: `employee_id` and the total sales corresponding to that employee. Call this view `trans_by_employee`.

In [69]:
sql = """SELECT employee_id,
             SUM(unit_price * quantity) AS total_sales
             FROM transactions 
             GROUP BY employee_id
             LIMIT 5;"""

pd.read_sql(sql, con=conn)

employee_id  total_sales
0            0     84500.50
1            1    106550.00
2            3     89877.25
3            9     95568.75
4           10    101340.75

In [70]:
sql = """SELECT * FROM trans_by_employee LIMIT 5;"""

pd.read_sql(sql, con=conn)

employee_id  total_sales
0            0     84500.50
1            1    106550.00
2            3     89877.25
3            9     95568.75
4           10    101340.75

## Common Table Expressions (CTEs)

67) What's the most common first initial for pets in the `pets` table?

In [71]:
sql = """WITH INIT(lower_case_init)
         AS 
         (SELECT LOWER(SUBSTR(name,1,1))
         FROM pets)
         
         SELECT lower_case_init,
                COUNT(lower_case_init) AS cnt
         FROM INIT
         GROUP BY lower_case_init
         ORDER BY cnt DESC
         LIMIT 1;"""

pd.read_sql(sql,con=conn)

lower_case_init  cnt
0               p    3

68) Create taglines for each employee in the `employees` table. As a template, the first row of the result should look like this:

```
Christine Thompson started in 2005 and makes $123,696 working in sales. 
```

To do this easily, make a CTE featuring name (firstname + " " + lastname), job, salary (formatted), and year. Job title should be lowercased, _unless_ it is IT, in which case leave it capitalized. The solution is simple string concatenation off of this long CTE.


In [72]:
sql = """WITH tagline(name,year,salary,job)
         AS
        (SELECT firstname || " " || lastname AS name,
            SUBSTR(startdate,1,4) AS year,
            PRINTF("$%,.2d",salary) AS salary,
            (CASE WHEN job == "IT" THEN "IT" ELSE LOWER(job) END) AS job
         FROM employees)
         
         SELECT name || " started in " \
            || year || " and makes " || \
            salary || " working in " || job || "." AS tagline
         FROM tagline LIMIT 5;"""

df = pd.read_sql(sql, con=conn)

In [73]:
list(df['tagline'])

['Christine Thompson started in 2005 and makes $123,696 working in sales.',
 'Thomas Peck started in 2011 and makes $112,972 working in sales.',
 'Christopher Robles started in 2003 and makes $78,426 working in IT.',
 'Elizabeth Munoz started in 1993 and makes $55,824 working in sales.',
 'Janice Martin started in 2011 and makes $62,007 working in IT.']

69) How many of our sales come from companies ending in each of "LLC", "Inc", "Ltd", or "PLC"? In a CTE, create a `company_type` column of values `"LLC"`, `"Inc"`, `"Ltd"`, `"PLC"`, or `"Other"`. Outside the CTE, find the total revenue from these categories, as well as their respective counts.
* _Hint:_ You'll need the `INSTR()` function.

In [74]:
sql = """WITH company_type(order_id,customer,unit_price,quantity,orderdate,employee_id,comp_type)
         AS 
         (SELECT *,
            (CASE WHEN INSTR(customer,"LLC") > 0 OR \
                INSTR(customer,"Inc") > 0 OR \
                INSTR(customer,"Ltd") > 0 OR \
                INSTR(customer,"PLC") > 0
                THEN SUBSTR(customer,-3,3) ELSE "Other" END) AS comp_type
         FROM transactions)
         
         SELECT comp_type,
             COUNT(*) AS cnt_company_type,
             SUM(unit_price * quantity) AS revenue
         FROM company_type
         GROUP BY comp_type;"""

pd.read_sql(sql, con=conn)

comp_type  cnt_company_type     revenue
0       Inc              4070   556110.75
1       LLC              2026   280666.75
2       Ltd              1577   215068.75
3     Other             39956  5459840.00
4       PLC              2452   341368.00

### Joins

70) Which employee made which sale? Join the `employees` table onto the `transactions` table by `employee_id`. You only need to include the employee's first/last name from `employees`.


In [75]:
sql = """SELECT t.*,
            e.firstname AS firstname,
            e.lastname AS lastname
         FROM transactions AS t 
         LEFT JOIN employees AS e
             ON t.employee_id = e.ID
        LIMIT 5;"""

pd.read_sql(sql, con=conn)

order_id                   customer  unit_price  quantity   orderdate  \
0         0  Kent, Curtis and Williams       20.50        12  2010-05-30   
1         0  Kent, Curtis and Williams       24.00        11  2010-05-30   
2         0  Kent, Curtis and Williams       22.25        14  2010-05-30   
3         0  Kent, Curtis and Williams       10.50        11  2010-05-30   
4         0  Kent, Curtis and Williams       13.75         9  2010-05-30   

   employee_id    firstname lastname  
0           81  Christopher  Carlson  
1           81  Christopher  Carlson  
2           81  Christopher  Carlson  
3           81  Christopher  Carlson  
4           81  Christopher  Carlson

71) What is the name of the employee who made the most in sales? Find this answer by doing a join as in the previous problem. Your resulting query will be difficult for someone else to read.

In [76]:
sql = """SELECT t.employee_id,
            SUM(t.unit_price * t.quantity) AS revenue,
            e.firstname AS firstname,
            e.lastname AS lastname
         FROM transactions AS t 
         LEFT JOIN employees AS e
             ON t.employee_id = e.ID
         GROUP BY employee_id
         ORDER BY revenue DESC
         LIMIT 1;"""

pd.read_sql(sql, con=conn)

employee_id    revenue firstname lastname
0           60  124372.25      Anna  Fischer

72) Solve the previous problem by joining `employees` onto the `trans_by_employee` view you made earlier.


73) Solve the previous problem by joining `employees` onto a CTE.
** Done in the previous exercise

74) Next, the company will try to give bonuses based on performance. Show all employees who've made more in sales than 1.5 times their salary. (You may use whatever technique you'd like to do the join: view, CTE, or even a subquery!)

In [77]:
sql = """WITH trans_by_employee(employee_id,total_sales)
            AS
            (SELECT employee_id,
             SUM(unit_price * quantity) AS total_sales
             FROM transactions 
             GROUP BY employee_id)
        
        SELECT t.employee_id,
               t.total_sales,
               e.firstname,
               e.lastname,
               s.salary
        FROM trans_by_employee AS t
        LEFT JOIN employees AS e
            ON t.employee_id = e.ID
        LEFT JOIN employees_null AS s
            ON t.employee_id = S.ID
        WHERE s.salary IS NOT NULL
        AND t.total_sales > 1.5 * s.salary
        ORDER BY t.total_sales DESC
        LIMIT 10;"""

pd.read_sql(sql, con=conn)

employee_id  total_sales firstname  lastname  salary
0           97    116507.50  Brittany   Maynard   69257
1           55    112284.25    Julian  Martinez   34597
2           73    111021.00    Miguel  Williams   31584
3           65    110207.00   Michael      West   32781
4           51    110153.75   William      Roth   55591
5           54    109555.25  Virginia       Liu   58582
6           33    109441.75   Shannon    Bailey   66071
7           19    107655.75     Roger    Conner   60375
8           13    105828.00  Veronica    Hughes   63788
9           44    104381.00      Lisa   Delgado   36704

75) Do we have potentially erroneous rows? Find all transactions which occurred _before_ the employee was even hired! (Make sure each transaction only occupies one row).


In [78]:
sql = """SELECT DISTINCT e.ID,
            e.firstname,
            e.lastname,
            e.startdate,
            t.order_id,
            t.orderdate
         FROM employees_null AS e
         LEFT JOIN transactions AS T
             on e.ID = t.employee_id
         WHERE orderdate < startdate
         LIMIT 10;"""

pd.read_sql(sql, con=conn)

ID  firstname  lastname   startdate  order_id   orderdate
0   0  Christine  Thompson  2005-01-20      1010  1993-02-20
1   0  Christine  Thompson  2005-01-20      1029  1991-12-20
2   0  Christine  Thompson  2005-01-20      1036  1997-08-20
3   0  Christine  Thompson  2005-01-20      1189  1998-06-08
4   0  Christine  Thompson  2005-01-20      1476  1993-11-20
5   0  Christine  Thompson  2005-01-20      1566  2000-03-19
6   0  Christine  Thompson  2005-01-20      1599  2004-01-01
7   0  Christine  Thompson  2005-01-20      1638  1999-06-14
8   0  Christine  Thompson  2005-01-20      2006  1991-03-19
9   0  Christine  Thompson  2005-01-20      2333  1993-07-21

76) Among all transactions that occurred from 2015 to 2019, create a table that is the monthly revenue of our company versus the total trading volume of Yum! in that month. Format the columns nicely. That is, a sample row of your result might look like this:

```
| year | month | company_revenue | yum_trade_volume |
|------|-------|-----------------|------------------|
| 2017 |    03 |        $100,000 |      125,000,000 |
```

* _Hint:_ You don't need any `WHERE` statements here. You can get the right answer simply by changing what kind of join you do!

77) Repeat the previous problem, but in addition to the total volume, include:
* The lowest price that month (ie, lowest low)
* The highest price that month (ie, highest high)

In [79]:
sql = """WITH revenue(month_year,company_revenue)
         AS (SELECT SUBSTR(orderdate,1,7) AS month_year,
            SUM(unit_price * quantity) AS company_revenue
         FROM transactions
         GROUP BY month_year),
        
         volume(month_year,yum_trade_volume,min_price,max_price)
         AS (SELECT SUBSTR(date,1,7) AS month_year,
            SUM(volume) AS yum_trade_volume,
            MIN(close) AS min_price,
            MAX(close) AS max_price
         FROM yum
         GROUP BY month_year)
         
         SELECT r.month_year,
                r.company_revenue,
                v.yum_trade_volume,
                v.min_price,
                v.max_price,
                SUBSTR(r.month_year,1,4) AS year
         FROM revenue AS r
         LEFT JOIN volume AS v
             ON r.month_year = v.month_year
         WHERE CAST(year AS int) >= 2015 AND CAST(year AS int) <= 2019
         LIMIT 12;"""

pd.read_sql(sql,con=conn)

month_year  company_revenue  yum_trade_volume  min_price  max_price  year
0     2015-01         15958.25          89074400  50.330696  53.393242  2015
1     2015-02         18113.50          98621800  52.286125  58.310570  2015
2     2015-03         21291.75         108827600  54.996407  58.806614  2015
3     2015-04         15088.25         117743300  56.283249  62.616821  2015
4     2015-05         16024.00         131485600  64.133720  68.209923  2015
5     2015-06         15712.50          69412500  64.421280  66.678650  2015
6     2015-07         18237.75         116786700  61.811646  66.132278  2015
7     2015-08         15764.00         155692800  54.708843  64.011505  2015
8     2015-09         13857.25          97615900  54.636951  59.719627  2015
9     2015-10         22777.75         232475400  48.497482  59.971245  2015
10    2015-11         20351.50          98766200  48.253056  52.451473  2015
11    2015-12         16791.50         116614800  51.258087  54.931705  2015

### Window Functions

78. Create a daily cumulative sum of the trading volume of the Yum! stock in yum.

In [80]:
sql = """SELECT date,
            volume,
            SUM(volume) OVER (ORDER BY date) AS sum_volume_by_day
         FROM yum
         LIMIT 20;"""

pd.read_sql(sql,con=conn)

date   volume  sum_volume_by_day
0   2015-01-02  2283400            2283400
1   2015-01-05  4418600            6702000
2   2015-01-06  5004400           11706400
3   2015-01-07  4554100           16260500
4   2015-01-08  4258200           20518700
5   2015-01-09  3687200           24205900
6   2015-01-12  3924400           28130300
7   2015-01-13  4612400           32742700
8   2015-01-14  5022200           37764900
9   2015-01-15  5393700           43158600
10  2015-01-16  4074600           47233200
11  2015-01-20  5162100           52395300
12  2015-01-21  3900900           56296200
13  2015-01-22  4975700           61271900
14  2015-01-23  4535900           65807800
15  2015-01-26  3141700           68949500
16  2015-01-27  4944000           73893500
17  2015-01-28  6663400           80556900
18  2015-01-29  4166800           84723700
19  2015-01-30  4350700           89074400

79. Create a cumulative sum of the trading volume of Yum! across months. That is, the final row of this query should be the cumulative sum of all months from 2015 through 2019.

In [81]:
sql = """SELECT DISTINCT(SUBSTR(date,1,7)) AS date_month,
            SUM(volume) OVER (ORDER BY SUBSTR(date,1,7)) AS sum_volume_by_month
         FROM yum
         LIMIT 5;"""

pd.read_sql(sql,con=conn)

date_month  sum_volume_by_month
0    2015-01             89074400
1    2015-02            187696200
2    2015-03            296523800
3    2015-04            414267100
4    2015-05            545752700

80. For March 2017, create a table from yum with the following columns:
    * Day of the month
    * Row number (ie, the nth trading day of the month)
    * Cumulative low (ie, lowest low so far this month)
    * Cumulative high (ie, highest high so far this month)
    * Cumulative total volume

In [82]:
sql = """SELECT date,
            low,
            high,
            volume,
            SUBSTR(date,9,2) AS day_of_month,
            SUM(1) OVER (ORDER BY SUBSTR(date,9,2)) AS row_nbr,
            MIN(low) OVER (ORDER BY SUBSTR(date,9,2)) AS cum_low,
            MAX(high) OVER (ORDER BY SUBSTR(date,9,2)) AS cum_high,
            SUM(volume) OVER (ORDER BY SUBSTR(date,9,2)) AS cum_volume
         FROM yum
         WHERE date BETWEEN "2017-03-01" AND "2017-03-31"
         LIMIT 5;"""

pd.read_sql(sql,con=conn)

date        low       high   volume day_of_month  row_nbr    cum_low  \
0  2017-03-01  65.370003  65.949997  2311500           01        1  65.370003   
1  2017-03-02  64.919998  65.500000  2176000           02        2  64.919998   
2  2017-03-03  64.419998  65.300003  2170400           03        3  64.419998   
3  2017-03-06  64.339996  64.720001  1985400           06        4  64.339996   
4  2017-03-07  63.990002  64.419998  2070900           07        5  63.990002   

    cum_high  cum_volume  
0  65.949997     2311500  
1  65.949997     4487500  
2  65.949997     6657900  
3  65.949997     8643300  
4  65.949997    10714200

84. From yum, create a column that represents the 7-day moving average of the closing price. This syntax is very confusing. You can read more about it here.

In [83]:
sql = """SELECT *,
            close,
            AVG(close) 
            OVER (ORDER BY date ROWS BETWEEN 6 PRECEDING AND CURRENT ROW) AS ma_close
         FROM yum
         LIMIT 14;"""

pd.read_sql(sql,con=conn)

date       open       high        low      close   volume  \
0   2015-01-02  52.796551  52.846874  51.761322  52.012939  2283400   
1   2015-01-05  52.056076  52.278934  50.948959  50.956146  4418600   
2   2015-01-06  51.301224  51.329979  49.877785  50.330696  5004400   
3   2015-01-07  51.020847  52.056076  50.884254  51.998562  4554100   
4   2015-01-08  52.408340  53.040977  52.365204  52.904385  4258200   
5   2015-01-09  52.990654  53.033787  52.322071  52.810928  3687200   
6   2015-01-12  52.588066  52.818115  52.242992  52.415527  3924400   
7   2015-01-13  52.933140  53.019409  51.689430  52.199856  4612400   
8   2015-01-14  51.265278  52.508987  51.171818  51.984184  5022200   
9   2015-01-15  51.861969  52.329258  51.337166  51.365925  5393700   
10  2015-01-16  51.279655  52.127964  50.984901  52.070454  4074600   
11  2015-01-20  52.063263  52.221424  51.006470  51.150253  5162100   
12  2015-01-21  50.891445  51.480949  50.762043  51.401871  3900900   
13  2015-01-22  51.761322  53.292595  51.538460  53.048168  4975700   

        close   ma_close  
0   52.012939  52.012939  
1   50.956146  51.484543  
2   50.330696  51.099927  
3   51.998562  51.324586  
4   52.904385  51.640546  
5   52.810928  51.835609  
6   52.415527  51.918455  
7   52.199856  51.945157  
8   51.984184  52.092020  
9   51.365925  52.239910  
10  52.070454  52.250180  
11  51.150253  51.999590  
12  51.401871  51.798296  
13  53.048168  51.888673

85. Repeat the March 2017 problem but instead of cumulative highs, lows, and totals, show the 5-day moving highs and lows. (No need for volume here.)

In [84]:
sql = """SELECT date,
            high,
            low,
            MAX(high) OVER (ORDER BY date ROWS BETWEEN 4 PRECEDING AND CURRENT ROW) as ma_high,
            MIN(low) OVER (ORDER BY date ROWS BETWEEN 4 PRECEDING AND CURRENT ROW) as ma_low
         FROM yum
         LIMIT 6;"""

pd.read_sql(sql,con=conn)

date       high        low    ma_high     ma_low
0  2015-01-02  52.846874  51.761322  52.846874  51.761322
1  2015-01-05  52.278934  50.948959  52.846874  50.948959
2  2015-01-06  51.329979  49.877785  52.846874  49.877785
3  2015-01-07  52.056076  50.884254  52.846874  49.877785
4  2015-01-08  53.040977  52.365204  53.040977  49.877785
5  2015-01-09  53.033787  52.322071  53.040977  49.877785

83. The Williams %R is an economic trendline indicator of a stock. Query yum to only include two columns: the date (unmodified), and the 7-day Williams %R of the stock at that date, call it williams_r. It is computed as follows:
    * Let h7 be the running 7-day high (ie, highest high of the last 7 trading days).
    * Let l7 be the running 7-day low (ie, lowest low of the lsat 7 trading days).
    * williams_r = (h7 - close) / (h7 - l7)

In [85]:
sql = """WITH williams(date,h7,l7,close)
         AS
         (SELECT date,
             MAX(high) OVER (ORDER BY date ROWS BETWEEN 6 PRECEDING AND CURRENT ROW) AS h7,
             MIN(low) OVER (ORDER BY date ROWS BETWEEN 6 PRECEDING AND CURRENT ROW) AS l7,
             close
          FROM yum)
         
         SELECT date,
             (h7-close) / (h7-l7) AS williams_r
         FROM williams
         LIMIT 10;
         """
pd.read_sql(sql,con=conn)

date  williams_r
0  2015-01-02    0.768213
1  2015-01-05    0.996213
2  2015-01-06    0.847458
3  2015-01-07    0.285715
4  2015-01-08    0.043182
5  2015-01-09    0.072727
6  2015-01-12    0.197727
7  2015-01-13    0.265909
8  2015-01-14    0.334091
9  2015-01-15    0.776666

84. Next, let's create the Stochastic Oscillator of yum. The stochastic oscillator is actually two lines: One called %K and the other called %D. They are computed as follows:
    * Let h14 and l14 denote the 14-day highs and lows (similar to last problem)
    * percent_k = (close - l14) / (h14 - l14)
    * percent_d is the 3-day moving average of percent_k

In [86]:
sql = """WITH values_hl14(date,l14,h14,close)
         AS
         (SELECT date,
            MIN(low) OVER (ORDER BY date ROWS BETWEEN 13 PRECEDING AND CURRENT ROW) AS l14,
            MAX(high) OVER (ORDER BY date ROWS BETWEEN 13 PRECEDING AND CURRENT ROW) AS h14,
            close
         FROM yum)
         
         SELECT *,
             (close - l14) / (h14 - l14) AS percent_k,
             AVG((close - l14) / (h14 - l14)) OVER (ORDER BY date ROWS BETWEEN 2 PRECEDING AND CURRENT ROW) AS percent_d
         FROM values_hl14
         LIMIT 5;"""

pd.read_sql(sql,con=conn)

date        l14        h14      close  percent_k  percent_d
0  2015-01-02  51.761322  52.846874  52.012939   0.231787   0.231787
1  2015-01-05  50.948959  52.846874  50.956146   0.003787   0.117787
2  2015-01-06  49.877785  52.846874  50.330696   0.152542   0.129372
3  2015-01-07  49.877785  52.846874  51.998562   0.714285   0.290205
4  2015-01-08  49.877785  53.040977  52.904385   0.956818   0.607882

In [87]:
sql = """SELECT * FROM yum LIMIT 5;"""

pd.read_sql(sql,con=conn)

date       open       high        low      close   volume
0  2015-01-02  52.796551  52.846874  51.761322  52.012939  2283400
1  2015-01-05  52.056076  52.278934  50.948959  50.956146  4418600
2  2015-01-06  51.301224  51.329979  49.877785  50.330696  5004400
3  2015-01-07  51.020847  52.056076  50.884254  51.998562  4554100
4  2015-01-08  52.408340  53.040977  52.365204  52.904385  4258200

In [88]:
sql = """SELECT t.*,
                y.*,
                SUBSTR(date,1,4) AS year,
                SUBSTR(date,6,2) AS month,
                SUM(unit_price * quantity) OVER (ORDER BY SUBSTR(date,1,7)) AS company_revenue,
                MIN(low) OVER (ORDER BY SUBSTR(date,1,7)) AS min,
                MAX(high) OVER (ORDER BY SUBSTR(date,1,7)) AS max,
                FIRST_VALUE(open) OVER (PARTITION BY SUBSTR(date,1,7) ORDER BY SUBSTR(date,1,7)) AS fv,
                LAST_VALUE(close) OVER (PARTITION BY SUBSTR(date,1,7) ORDER BY SUBSTR(date,1,7)) AS cv,
                SUM(volume) OVER (ORDER BY SUBSTR(date,1,7)) AS tot_volume
            FROM transactions AS t
            LEFT JOIN yum AS y
                ON t.orderdate = y.date
            WHERE y.date >= "2015-01-01"
            LIMIT 5;"""

pd.read_sql(sql,con=conn)

order_id      customer  unit_price  quantity   orderdate  employee_id  \
0      1993  Lopez-Fuller        7.25        11  2015-01-02           25   
1      1993  Lopez-Fuller       13.00        15  2015-01-02           25   
2      1993  Lopez-Fuller       20.75         7  2015-01-02           25   
3      9607   Evans-Parks       12.00        10  2015-01-06           66   
4      9607   Evans-Parks       19.25         9  2015-01-06           66   

         date       open       high        low      close   volume  year  \
0  2015-01-02  52.796551  52.846874  51.761322  52.012939  2283400  2015   
1  2015-01-02  52.796551  52.846874  51.761322  52.012939  2283400  2015   
2  2015-01-02  52.796551  52.846874  51.761322  52.012939  2283400  2015   
3  2015-01-06  51.301224  51.329979  49.877785  50.330696  5004400  2015   
4  2015-01-06  51.301224  51.329979  49.877785  50.330696  5004400  2015   

  month  company_revenue        min       max         fv         cv  \
0    01           8859.5  49.877785  53.41481  52.796551  51.962616   
1    01           8859.5  49.877785  53.41481  52.796551  51.962616   
2    01           8859.5  49.877785  53.41481  52.796551  51.962616   
3    01           8859.5  49.877785  53.41481  52.796551  51.962616   
4    01           8859.5  49.877785  53.41481  52.796551  51.962616   

   tot_volume  
0   303041100  
1   303041100  
2   303041100  
3   303041100  
4   303041100